### DISCUSSION TITLE
why need to add route in net2net-ip6-in-ip4-ikev2?

### DISCUSSION THREAD

**Discussion by ww20** [strongswan/strongswan#595] [Q&A]
Hi,

I am trying to set up following strongswan scenario in our system environment. 

> https://www.strongswan.org/testing/testresults/ipv6/net2net-ip6-in-ip4-ikev2/

In its console.log, there are some commands as below to add routes. Can you explain why we need to add these routes? Will swanctl.conf take care of the routing below? And why there is no adding routes in the console.log of the net2net-route scenario (https://www.strongswan.org/testing/testresults/ikev2/net2net-route/)?

> alice# ip route add fec2::/16 via fec1::1
> moon# ip route add fec2::/16 via fec0::2
> sun# ip route add fec1::/16 via fec0::1
> bob# ip route add fec1::/16 via fec2::1

Thanks.

**Comment by Thermi** [comment]
Because the kernel needs a next hop to route the packets to. Otherwise sending/forwarding of the IPv6 packets to the remote subnet will fail with `EHOSTUNREACH` (No route to destination).

**Comment by tobiasbrunner** [comment]
Manually adding the routes on _sun_ and _moon_ would not be required, but for some reason, route installation by strongSwan is disabled in that particular test scenario (via `charon.install_routes`). Don't know why, could be that there were some issues in earlier strongSwan or kernel versions with IPv6 routing (or specifically for the IPv6 in IPv4 case).

The routes on _alice_ and _bob_ are required if they don't already have a route via VPN gateway for the remote networks (see [ForwardingAndSplitTunneling](https://wiki.strongswan.org/projects/strongswan/wiki/ForwardingAndSplitTunneling)), which isn't the case here for these remote IPv6 subnets.

**Comment by ww20** [comment]
I have been trying to set up [net2net-ip6-in-ip4-ikev2](https://www.strongswan.org/testing/testresults/ipv6/net2net-ip6-in-ip4-ikev2/)  in our environment but it did not work so far. The issue is, after successfully running `swanctl --initiate --child net-net` on both moon and sun, I cannot ping `fec2::1` on sun or `fec2::10` on bob from moon or alice. Ping didn't work either after enabling charon.install_routes (some route was added in table 220 though). 

`fec0::1` and `fec0::2` are not specified in the example diagram. Is `fec0::2` on eth0 of sun and `fec0::1` on eth0 of moon? If so, will ping command work if we don't have ipv6 address `fec0::1` and `fec0::2` available in our environment(no routes below)? We do have ipv4 address on eth0 though. Any another reason why ping does not work?

```
moon# ip route add fec2::/16 via fec0::2
sun# ip route add fec1::/16 via fec0::1
```

```
moon>ping6 <ipv6 address in fec2 subnet>
......
From <ipv6 address in fec1 subnet> icmp_seq=1 Destination unreachable: Address unreachable
From <ipv6 address in fec1 subnet> icmp_seq=2 Destination unreachable: Address unreachable
From <ipv6 address in fec1 subnet> icmp_seq=3 Destination unreachable: Address unreachable
```


Thanks.

**Reply by Thermi** [reply]
What is `<fec2 address>` exactly? Please make sure there are no typos.

**Reply by ww20** [reply]
I updated `<fec2 address>` to `<ipv6 address in fec2 subnet>`. "fec2" is labeled in the diagram from [net2net-ip6-in-ip4-ikev2](https://www.strongswan.org/testing/testresults/ipv6/net2net-ip6-in-ip4-ikev2/). Let me know if it is still not clear.

**Reply by tobiasbrunner** [reply]
Why mask the address at all?

**Reply by ww20** [reply]
I thought that might be helpful to understand the issue by using example address.

```
> ping6 fdfe:14db:ffac:0:201:4449:fc9b:7cb2
PING fdfe:14db:ffac:0:201:4449:fc9b:7cb2(fdfe:14db:ffac:0:201:4449:fc9b:7cb2) 56 data bytes
From fd12:1a90:57a0:0:201:444d:a7a7:55ea icmp_seq=1 Destination unreachable: Address unreachable
From fd12:1a90:57a0:0:201:444d:a7a7:55ea icmp_seq=2 Destination unreachable: Address unreachable
From fd12:1a90:57a0:0:201:444d:a7a7:55ea icmp_seq=3 Destination unreachable: Address unreachable
```

**Reply by tobiasbrunner** [reply]
It's not and neither is this without providing all the context information that goes along with it (routes, addresses, configs, status output, firewall rules etc. see #196 - without masking any of it!).

**Reply by ww20** [reply]
We are not sure if we can provide all of the information. However, we have a question in terms of the example [net2net-ip6-in-ip4-ikev2](https://www.strongswan.org/testing/testresults/ipv6/net2net-ip6-in-ip4-ikev2/) you provide. Essentially, what we want to know is where does the address fec0::2 reside? Basically on what interface and node?

`moon# ip route add fec2::/16 via fec0::2`

**Reply by tobiasbrunner** [reply]
It's on sun's `eth0`.

**Reply by ww20** [reply]
Is the implication that, in addition to the ipv4 addresses in the example, a valid site local address (now unique local) is required on both moon and sun. Where moon can contact sun's address and vice versa? Which acts as the "next hop". Thank you.

**Reply by Thermi** [reply]
Look. Using XFRM does not magically make the necessity of routes go away. You still need a valid route to the destination subnet. If you do not have any logical Ipv6 connectivity, you at least need, for example, a dummy device (never tested that theory) over which you at least have a route to the remote subnet. Preferably, a complete default route. That way charon doesn't need to insert any additional routes. Otherwise route selection fails exactly how you saw it: `Destination unrheachable: Address unreachable`.

XFRM steals away the packet before it's actually sent, and routes the ESP (in (udp|tcp) packet itself again, regardless of the original destination. But for that all to occur first, the route selection by the routing engine first needs to succeed. Even if it has a bogus next hop or output device. At the very least then, the selected output device needs to have disable_policy and disable_xfrm set to `0` (default value).

In your case, you don't have a route to fdfe:[...], so sending of the packet fails.

EDIT: replaed remote IKe responder with remote subnet because that's correct. It makes no sense to have an IPv6 route to an IPv4 destination IP address (unless you use transitional technology that uses that, but you don't do that here).

**Reply by ww20** [reply]
The dummy route approach works on our side. Thank you!
